### PR TITLE
Bluetooth - proper state machine implemented, including one for call management

### PR DESCRIPTION
**Pull request by pholat**
Bluetooth state machine merged with wonky switch case. Now there is one clear way to handle events between os and Bluetooth driver. Includes state machine for calls - which should fix Bluetooth call issues.

Make sure that this PR:
- [x] Complies with our guidelines for contributions
- [x[ Has unit tests if possible.
- [x] Has documentation updated

### CODE REVIEW

File: module-bluetooth/Bluetooth/BluetoothWorker.cpp [hunk 1]
```diff
@@ -141,33 +155,18 @@ auto BluetoothWorker::run() -> bool
...
-        controller->processCommand(command);
-        controller->turnOff();
-        break;
-    default:
-        controller->processCommand(command);
-        break;
+    LOG_INFO("handle bluetooth command(s)");
+    xQueueReceive(queue, nullptr, 0);
+    while (not workerQueue->empty()) {
+        auto cmd = workerQueue->peek();
```
Comment (kkleczkowski) [inline module-bluetooth/Bluetooth/BluetoothWorker.cpp:161]: Pleas use full words: cmd ->command, evt->event

**Review by kkleczkowski — APPROVED** [review]

**Review by SP2FET — COMMENTED** [review]
comments from mob review

File: module-bluetooth/Bluetooth/BluetoothStateMachine.hpp [hunk 2]
```diff
@@ -0,0 +1,442 @@
+// Copyright (c) 2017-2022, Mudita Sp. z.o.o. All rights reserved.
+// For licensing, see https://github.com/mudita/MuditaOS/LICENSE.md
+
+#pragma once
+
+#define BOOST_SML_CFG_DISABLE_MIN_SIZE // GCC10 fix
```
[the remark this answers is not in the document]
Comment (pholat) [reply module-bluetooth/Bluetooth/BluetoothStateMachine.hpp:6]: I did'nt want to touch it as I do not know why it was added in the first place.
As I've checked locally - it didn't change anything to me either in size, nor in functionality :)

File: module-services/service-bluetooth/ServiceBluetooth.cpp [hunk 3]
```diff
@@ -209,7 +193,7 @@ auto ServiceBluetooth::handle(message::bluetooth::SetStatus *msg) -> std::shared
     case BluetoothStatus::State::On:
 
         cpuSentinel->HoldMinimumFrequency(bsp::CpuFrequencyMHz::Level_3);
-        sendWorkerCommand(bluetooth::Command::Type::PowerOn);
+        sendWorkerCommand(new bt::evt::PowerOn());
```
[the remark this answers is not in the document]
Comment (pholat) [reply module-services/service-bluetooth/ServiceBluetooth.cpp:196]: in previous handling, we were using unique_ptr as raw ptr with just a bit more code than usual :) Then I've added a queue to receive messages nicely and now we are able to properly handle it and I simply forgot to fix it fully. Will be added

File: module-bluetooth/Bluetooth/BluetoothStateMachine.hpp [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+                                             state<Setup>[not isInit] = "Restart"_s,
+                                             state<Setup> + exception<InitializationError> / printInitError = "Off"_s,
+
+                                             state<On> + event<bt::evt::PowerOff> / TurnOff = "Off"_s,
+                                             state<On> + exception<ProcessingError> / ( printProcessingError, TurnOff ) = "Restart"_s,
+                                             state<On> + event<bt::evt::ShutDown> / TurnOff = X,
+
+                                             "Restart"_s = state<Setup>,
+                                             "Restart"_s + event<bt::evt::ShutDown> /TurnOff = X,
+
```
[the remark this answers is not in the document]
Comment (pholat) [reply module-bluetooth/Bluetooth/BluetoothStateMachine.hpp:411]: I believe we should add a task for it. I've double-checked that logic here is not changed and no regression is added, but logic seems a bit lacking here.

File: module-bluetooth/Bluetooth/BluetoothWorker.cpp [hunk 5]
```diff
@@ -141,33 +155,18 @@ auto BluetoothWorker::run() -> bool
...
+        if (cmd == std::nullopt) {
+            LOG_ERROR("There was no data even with notification");
+            break;
+        }
+        if ((*cmd).evt != nullptr) {
+            controller->handle(*cmd->evt);
+            delete cmd->evt;
+        }
```
[the remark this answers is not in the document]
Comment (pholat) [reply module-bluetooth/Bluetooth/BluetoothWorker.cpp:169]: will be fixed :)

File: module-bluetooth/Bluetooth/BluetoothStateMachine.hpp [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+                        state<Idle> + event<bt::evt::BatteryLevelData>/ SetBatteryLevel = state<Idle>,
+                        state<Idle> + event<bt::evt::NetworkStatusData> / SetNetworkStatus = state<Idle>,
+                        state<Idle> + event<bt::evt::StartStream>/ StartAudio = state<Idle>,
+                        state<Idle> + event<bt::evt::StopStream>/ StopAudio = state<Idle>,
+
+                        state<Idle> + event<bt::evt::StartRouting> / forwardEvent= state<Call>,
+                        state<Idle> + event<bt::evt::StartRinging> / forwardEvent= state<Call>,
+                        state<Idle> + event<bt::evt::IncomingCallNumber>  / forwardEvent  = state<Call>,
+                        state<Idle> + event<bt::evt::CallStarted> / forwardEvent= state<Call>,
+                        state<Call> = state<Idle> // this one is needed to go out from Call substate properly!
```
Comment (SP2FET) [inline module-bluetooth/Bluetooth/BluetoothStateMachine.hpp:364]:
```suggestion
                        state<Idle> + event<bt::evt::StartRouting> / forwardEvent= state<Call>,
                        state<Idle> + event<bt::evt::IncomingCallNumber>  / forwardEvent  = state<Call>,
                        state<Idle> + event<bt::evt::CallStarted> / forwardEvent= state<Call>,
                        state<Call> = state<Idle> // this one is needed to go out from Call substate properly!
```

**Review by SP2FET — APPROVED** [review]

File: module-bluetooth/Bluetooth/AbstractController.hpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2017-2022, Mudita Sp. z.o.o. All rights reserved.
+// For licensing, see https://github.com/mudita/MuditaOS/LICENSE.md
+
+#pragma once
+
+namespace bt::evt
```
[the remark this answers is not in the document]
Comment (pholat) [reply module-bluetooth/Bluetooth/AbstractController.hpp]: done

**Pull request merged**